internal/database/migration/definition/read.go

```diff
@@ -157,20 +169,217 @@ func hydrateDefinitions(fs fs.FS, definitions []Definition) (err error) {
+// of the file for extraction.
+func extractMetadata(contents string) (_ string, metadata Metadata, _ error) {
+	match := metadataPattern.FindStringSubmatch(contents)
+
+	if len(match) > 0 {
+		if err := yaml.Unmarshal([]byte(extractCommentPrefix(match[1])), &metadata); err != nil {
+			return "", metadata, err
+		}
+
+		return strings.TrimSpace(contents[len(match[0]):]), metadata, nil
+	}
+
+	return contents, metadata, nil
+}
+
+// extractCommentPrefix removes the `-- ` on each line of the given multiline string.
+func extractCommentPrefix(match string) string {
+	lines := strings.Split(match, "\n")
+	for i := range lines {
+		lines[i] = strings.TrimPrefix(lines[i], "-- ")
+	}
+
+	return strings.Join(lines, "\n")
+}
+
+// reorderDefinitions will re-order the given migration definitions in-place so that
+// migrations occur before their dependents in the slice. An error is returned if the
+// given migration definitions do not form a single-root directed acyclic graph.
+func reorderDefinitions(migrationDefinitions []Definition) error {
+	if len(migrationDefinitions) == 0 {
+		return nil
+	}
+
+	// Stash migration definitions by identifier
+	migrationDefinitionMap := make(map[int]Definition, len(migrationDefinitions))
+	for _, migrationDefinition := range migrationDefinitions {
+		migrationDefinitionMap[migrationDefinition.ID] = migrationDefinition
+	}
+
+	// Find topological order of migrations
+	order, err := findDefinitionOrder(migrationDefinitions)
+	if err != nil {
+		return err
+	}
+
+	for i, id := range order {
+		// Re-order migration definitions slice to be in topological order. The order
+		// returned by findDefinitionOrder is reversed; we want parents _before_ their
+		// dependencies, so we fill this slice in backwards.
+		migrationDefinitions[len(migrationDefinitions)-1-i] = migrationDefinitionMap[id]
+	}
+
+	return nil
+}
+
+var (
+	ErrNoRoots       = fmt.Errorf("no roots")
+	ErrMultipleRoots = fmt.Errorf("multiple roots")
+	ErrCycle         = fmt.Errorf("cycle")
+)
+
+// findDefinitionOrder returns an order of migration definition identifiers such that
+// migrations occur only after their dependencies (parents). This assumes that the set
+// of definitions provided form a single-root directed acyclic graph and fails with an
+// error if this is not the case.
+func findDefinitionOrder(migrationDefinitions []Definition) ([]int, error) {
+	root, err := root(migrationDefinitions)
 	if err != nil {
 		return nil, err
 	}
 
-	normalizedQuery := string(contents)
-	normalizedQuery = strings.ReplaceAll(normalizedQuery, "%", "%%")
+	var (
+		order    = make([]int, 0, len(migrationDefinitions))
+		marks    = make(map[int]int, len(migrationDefinitions))
+		children = children(migrationDefinitions)
+
+		dfs func(id int) error
+	)
+
+	dfs = func(id int) error {
+		switch marks[id] {
+		case 1:
+			// currently processing
+			return ErrCycle
+		case 2:
+			// already visited
+			return nil
+		}
+
+		// temporary mark
+		marks[id]++
```

Comment:
What do you think about using explicit values here? Helps a bit with readability

```suggestion
		marks[id] = 1
```

Reply:
that would break the case where we check for 2 above, no?

Reply:
I implicitly also meant that the other `marks[id]++` further down gets turned into `marks[id] = 2` :)